src/ApiService/ApiService/onefuzzlib/NodeOperations.cs

```diff
@@ -20,4 +48,173 @@ public NodeOperations(IStorage storage, ILogTracer log, IServiceConfig config)
+            queryParts.Add($"delete_requested eq false");
+        }
+
+        //# azure table query always return false when the column does not exist
+        //# We write the query this way to allow us to get the nodes where the
+        //# version is not defined as well as the nodes with a mismatched version
+        var versionQuery = $"not (version eq '{oneFuzzVersion}')";
+        queryParts.Add(versionQuery);
+
+        return Query.And(queryParts);
+    }
+
+
+    public IAsyncEnumerable<Node> SearchStates(
+        Guid? poolId = default,
+        Guid? scaleSetId = default,
+        IList<NodeState>? states = default,
+        string? poolName = default,
+        bool excludeUpdateScheduled = false,
+        int? numResults = default) {
+        var query = NodeOperations.SearchStatesQuery(_config.OneFuzzVersion, poolId, scaleSetId, states, poolName, excludeUpdateScheduled, numResults);
+        return QueryAsync(query);
+    }
+
+    public async Async.Task MarkTasksStoppedEarly(Node node, Error? error = null) {
+        if (error is null) {
+            error = new Error(ErrorCode.TASK_FAILED, new[] { $"node reimaged during task execution.  machine_id: {node.MachineId}" });
+        }
+
+        await foreach (var entry in _nodeTasksOps.GetByMachineId(node.MachineId)) {
+            var task = await _taskOps.GetByTaskId(entry.TaskId);
+            if (task is not null) {
+                await _taskOps.MarkFailed(task, error);
+            }
+            if (!node.DebugKeepNode) {
+                await Delete(node);
+            }
+        }
+    }
+
+    public new async Async.Task Delete(Node node) {
+        await MarkTasksStoppedEarly(node);
+        await _nodeTasksOps.ClearByMachineId(node.MachineId);
+        await _nodeMessageOps.ClearMessages(node.MachineId);
+        await base.Delete(node);
+
+        await _events.SendEvent(new EventNodeDeleted(node.MachineId, node.ScalesetId, node.PoolName));
+    }
+}
+
+
+public interface INodeTasksOperations : IStatefulOrm<NodeTasks, NodeTaskState> {
+    IAsyncEnumerable<Node> GetNodesByTaskId(Guid taskId, INodeOperations nodeOps);
+    IAsyncEnumerable<NodeAssignment> GetNodeAssignments(Guid taskId, INodeOperations nodeOps);
+    IAsyncEnumerable<NodeTasks> GetByMachineId(Guid machineId);
+    IAsyncEnumerable<NodeTasks> GetByTaskId(Guid taskId);
+    Async.Task ClearByMachineId(Guid machineId);
+}
+
+public class NodeTasksOperations : StatefulOrm<NodeTasks, NodeTaskState>, INodeTasksOperations {
+
+    ILogTracer _log;
+
+    public NodeTasksOperations(IStorage storage, ILogTracer log, IServiceConfig config)
+        : base(storage, log, config) {
+        _log = log;
+    }
+
+    public async IAsyncEnumerable<Node> GetNodesByTaskId(Guid taskId, INodeOperations nodeOps) {
+        List<Node> results = new();
+        await foreach (var entry in QueryAsync($"task_id eq '{taskId}'")) {
+            var node = await nodeOps.GetByMachineId(entry.MachineId);
```

Comment:
this can probably be optimize by query all NodesTasks  then query the all machine in single request

Reply:
I'll add this suggestion as TODO in the code.

For now just porting logic 1 to 1 with python

Reply:
@stishkin this is what is meant 

``` C#
var nodetasks = await  QueryAsync($"task_id eq '{taskId}'").Select(x => x.MachineId.ToString()).ToListAsync();
var query = Query.EqualAny("RowKey", nodetasks);
await foreach( var node in nodeOps.QueryAsync(query))
{
    yield return node;
}
```